[WIP] Re implement de-multiplexing #825

hishnash opened the pull request:
This brings a version of de-multiplexing (for json websockets only) to v2. https://github.com/django/channels/issues/825

this adds 3 classes:

1) `class ApplicationWithChildren:`

(maybe we should find a better name)

this class managed the child Application coroutiens along side the Task/Dispatch for both the upstream and downstream queues for messages. 

2) `class Demultiplexer(ApplicationWithChildren):`

the base class for a `Demultiplexer`, compliant with ASGI application spec.

3) `class AsyncJsonWebsocketDemultiplexer(Demultiplexer, AsyncJsonWebsocketConsumer):`

this works as it used to (more or less, it now supports any valid JSON values for the `payload` since messages are in v2 passed within the same python thread so fewer limitations than before).

and example usage
```python
class MyDemultiplexer(AsyncJsonWebsocketDemultiplexer):
    applications = {
        "user": UserConsumer, 
        "posts": URLRouter([
            url("^/admin/$", AdminPostsConsumer),
            url("^/public/$", PublicPostsConsumer),
        ])
    }
```
---

There is an issue with the cleanup of tasks, https://github.com/hishnash/channels/blob/Re-Implement-Multiplexing-%23825/channels/utils.py#L54-L61  tasks are being cancelled even though the run loop is already closed. 

---

I still need to add Documentation before this is merged just wanted to check that this is a good approach to the problem before it put the time in for those bits.

andrewgodwin commented:
My first reaction to this is that it definitely shouldn't be a mixin - please make it just a normal class! Just reimplement the few bits you need from other classes.

hishnash commented:
ok sure, so just the `AsyncJsonWebsocketDemultiplexer` class.

andrewgodwin commented on channels/generic/websocket.py:
```diff
@@ -254,3 +256,155 @@ class AsyncJsonWebsocketConsumer(AsyncWebsocketConsumer):
+        """
+        child_tasks = []
+        try:
+            self.child_upstream_queues, child_tasks = await self._init_child_applications()
```
There's no need to make `init_child_applications` a separate method here really, it just makes it harder to read.

andrewgodwin commented on channels/generic/websocket.py:
```diff
@@ -254,3 +256,155 @@ class AsyncJsonWebsocketConsumer(AsyncWebsocketConsumer):
+        will be called otherwise message will be passed directly up the
+        application stack.
+        """
+        handler = getattr(self, get_handler_name(message), None)
```
Why do you need to intercept it like this? This class is only for websockets, you only need to handle three types of message (accept, send, close).

hishnash replied:
I see the use case where users might want to send from the child applications to the de-multiplexer. Since you need to subclass the de-multiplexer to use it having it like this will mean you just need to create the methods you want to call, like any other part of channels.

andrewgodwin commented on channels/generic/websocket.py:
```diff
@@ -254,3 +256,155 @@ class AsyncJsonWebsocketConsumer(AsyncWebsocketConsumer):
+        else:
+            raise ValueError("Invalid multiplexed **frame received (no channel/payload key)")
+
+    async def websocket_send(self, message, child_lookup):
```
Can we call `child_lookup` -> `stream_name` everywhere? That would make it clearer.

hishnash commented:
@andrewgodwin so with respect to `connect` and `disconnect` what i would suggest doing is a bit like v1:

a) the de-multiplexer accepts the connection all on its own. 
https://github.com/django/channels/blob/1.x/channels/generic/websockets.py#L273-L278

b) it calls `connect` to each child once those children accept they are added to the collection of children that can have messages routed to them.

c) if a disconnect comes through all that have responded `accept` will get the message.

c) if they trigger a disconnect they are removed from the set of child_applications and the `asyncio.Task` for them is cancelled.

d) when the last child disconnects (there are no more child applications left open) the de-multiplexer sends the disconnect onwards down.

in v1 the connection was accepted at the de-multiplexer, but if any child app sends a disconnect command it would disconnect the entire de-multiplexer not just disconnect that child from the de-multiplexer.

andrewgodwin commented:
Hmm, that seems reasonable enough, and people can override and subclass the Multiplexer if they want to put access control/connection denial on that.

ASGI applications should exit their main routine (and thus complete their Future) once they close their connection, so you should wait for all the applications to exit, rather than trying to work out if they have run their `disconnect` function (remember, you only get to send to them, you can't see what they're doing). It's likely worth doing what Daphne does and giving them a 10 second grace period to exit when the socket itself disconnects and they've been sent a disconnect message before just calling `task.cancel()` on them and exiting messily (we should raise a warning if we have to do this). You can see the relevant Daphne code here: https://github.com/django/daphne/blob/f53eb0dda6161c9fca78c8a89a1d94c7b613b1d4/daphne/server.py#L210

hishnash commented:
ok, given that the de-Multiplexer is already subject to that 10-second timeout should it not be a little less than the Daphne timeout otherwise we will never get the warning from the de-Multiplexer since the de-Multiplexer routine will timeout (10 seconds) before its internal timeout for its children. 

Also if any child raises an exception at any point (that is not StopConsumer) we should kill them, all right?

andrewgodwin commented:
Hm. You can't rely on daphne's timeout being 10 seconds (it's configurable), so you need to write a try/finally block that makes sure all the children are killed as the multiplexer's main coroutine shuts down, and maybe warn then (though daphne's warning might be enough to let the developer know it's not going well).

The multiplexer's `disconnect` can just run in parallel with the childrens' - there's no longer a socket there to talk to at that point, so it's not like they can do anything to the multiplexer.

If there is an exception, then yes, I think it is fair to drop the entire multiplexer and all children. The alternative is logging the traceback and continuing on, but I think hard failure here is correct.

hishnash commented:
> The multiplexer's disconnect can just run in parallel with the children's - there's no longer a socket there to talk to at that point, so it's not like they can do anything to the multiplexer.

it might still be nice to have that method be called after since it means all child applications are finished. Maybe you want to do some cleanup etc that needs to wait until all children are finished.

hishnash commented:
@andrewgodwin this is making `/generic/websocket.py` explode in size I'm going to break this out into a `mutliplexing.py` file.

hishnash commented:
with the `streams` possibly removing themselves does it make sense to raise a `ValueError` when the `stream` is not listed? does raising a `ValueError` close the socket?

hishnash commented:
@andrewgodwin I believe I have addressed your points with respect to connections and disconnections if you have time would you be able to give this solution a read over.

in the end it does feel rather convoluted so if you have an idea on how we can exit on Exception (when needed) and wait for the child to close (with a timeout for brute force killing) that does not require all of these extra features observing the child applications for errors.

also, what is the expected behaviour if an exception is raised. (in v1 this did not close the socket) but am i correct that in v2 this will result in the socket being closed? if so then maybe we should not reais exceptions when the `stream_name` cant be found but rather respond with an error frame?

andrewgodwin commented on channels/generic/multiplexer.py:
```diff
@@ -0,0 +1,322 @@
+
+        try:
+            # wait for all the upstream applications to close down
+            await asyncio.wait_for(
```
There's no need to wait for them to be killed. Just block forever and let the outside server kill your whole coroutine if things go bad.

hishnash replied:
then we wont be able to raise any warnings at this level we will just get the warning from daphne?

hishnash replied:
also will this not just result in there always being a warning being raised?

andrewgodwin commented on channels/generic/multiplexer.py:
```diff
@@ -0,0 +1,322 @@
+                for task in stream_tasks.values():
+                    task.cancel()
+
+    async def _monitor_task(self, stream_name: str, task: asyncio.Task):
```
StopConsumer should not bubble up out of other consumers (their coroutines will catch it), so I don't know why you have this method? Did you see it happen?

hishnash replied:
don't we still need this method so that if just one child application dies (without an exception) we remove it from the set of upstream streams but keep the rest running.  

if we just gather/waiton the coroutines without some wrapper method we will not be able to remove them from the set of open upstream.

hishnash replied:
don't we still need this method so that if just one child application dies (without an exception) we remove it from the set of upstream streams but keep the rest running?  

if we just gather/wait on the coroutines without some wrapper method we will not be able to remove them from the set of open upstream.

hishnash commented:
> This is still looking way too complicated in general 

I agree, it ran away from itself.

> I'd suggest rewriting it without basing it on a consumer class at all

ok, so what about having a `Demultiplex` class that takes a `multiplexer` class that would be a `JsonWebsocketMultiplexer` in this example (to do the needed web socket specifics). To split out that logic from the raw Demultiplex (handling of upstream applications etc)

hishnash commented:
@andrewgodwin I have made a lot of changes, I have tried to break out the logic a bit as well. 

however, I still think on `webseoct.disconect` we need to wait for the child applications to close not just block expecting the server to kill us otherwise we will always get a warning.

I moved all of the Queue observation (and dispatch) to a new parent class that is there to let one monitor both a collection of child application co-routines and the upstream/downstream queues and close them all in the correct manner regardless of who raises an exception.

hishnash commented:
I completely understand It had run away on itself if you can find a way to manage these child apps without all that complexity that is a much better solution. feel free to close this pull request.

I chose to split out into separate classes just to avoid having a supermassive 500 to 1000 line class to handle all the child task logic.

the big issues I had were:

- ensuring the tasks that wait for the upstream and downstream `Queues` are closed before the run loop is closed. (otherwise, there are loads and loads of warnings raised)
   - this is both the upstream and downstream message `Queues` on the de-multiplexer
   - and the upstream/layer message `Queues` for the child applications.
what would happen is if one child died (eg raised an exception) and I just let the bubble up then the full run-loop would be killed by the server (as it should) but some of the other tasks (as described above) would not be `cancelled` before this.

In addition, I believe it is good practice in this case not to close the downstream connection until all upstream tasks have closed. 

I'm rather new to the `async` `await` world in python so there may well be a much better simpler approach than those that I tried.

hishnash commented:
Im closing this for know

hishnash commented:
The changes in https://github.com/django/channels/commit/d6643d96dd59e153cda13681a6d5a8963b9d577c might well have solved the issues I had the resulted in such a complex solution....
